test/memcached/cache_impl_test.go

```diff
@@ -0,0 +1,526 @@
+// Adapted from test/redis/cache_impl_test.go, with most test cases being the same
+// basic idea. TestMemcacheAdd() is unique to the memcache tests, since redis can create a new key
+// simply by incrementing it but memcached cannot. In memcache new keys need to be explicitly
+// added.
+package memcached_test
+
+import (
+	"math/rand"
+	"strconv"
+	"testing"
+
+	"github.com/bradfitz/gomemcache/memcache"
+	"github.com/coocood/freecache"
+
+	pb "github.com/envoyproxy/go-control-plane/envoy/service/ratelimit/v3"
+	"github.com/envoyproxy/ratelimit/src/config"
+	"github.com/envoyproxy/ratelimit/src/limiter"
+	"github.com/envoyproxy/ratelimit/src/memcached"
+	"github.com/envoyproxy/ratelimit/src/utils"
+	stats "github.com/lyft/gostats"
+
+	"github.com/envoyproxy/ratelimit/test/common"
+	mock_memcached "github.com/envoyproxy/ratelimit/test/mocks/memcached"
+	mock_utils "github.com/envoyproxy/ratelimit/test/mocks/utils"
+	"github.com/golang/mock/gomock"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMemcached(t *testing.T) {
```

Comment:
as a matter of preference i would have split that test in 3 smaller ones

Reply:
This is intended to exactly mirror the test cases in the redis version. I wanted to keep it similar so there's hope of potentially extracting the test cases into some common location to share between redis and memcache, although I'm still not sure exactly what that looks like.